### ISSUE TITLE
Emacs rainbow delimiters mode support.

### ISSUE THREAD

**Issue by EskoJTH** [agda/agda#5015]
Would It be possible to allow rainbow-delimiters mode to color parenthesis? Currenlty agda is somehow overwriting its behaviour. Would help me to see different levels of nesting and I'm quite used to it being everywhere. I tried looking at the elisp code to do this, but I feel a bit too incompetent with elisp still to do this. 

I think It would be also very good if the coloring in general was bound to the overall user selected theme. I found the contrast with the default dark mode to be way too low and it took me very long selecting colors.

**Comment by jespercockx** [comment]
Welcome @EskoJTH and thanks for the suggestion. However you can already do what you're asking (by installing the right Emacs package and adding the setup to your `.emacs` file). Emacs is a very configurable platform and each person will have their own preference on how to set it up, so I don't think it's a very good idea to bundle this kind of functionality with Agda.

Regarding the color scheme, the Emacs mode currently automatically selects from two color schemes depending on whether a light or dark theme is selected. Do you suggest adding more color schemes (in which case, when should they be activated?) or do you just suggest improving the contrast of the current dark theme?

**Comment by ajrouvoet** [comment]
> I think It would be also very good if the coloring in general was bound to the overall user selected theme. 

I think this is indeed a problem with the current setup. Agda uses non-standard classes of syntax where it could make use of the usual syntactic classes. If it does, then the user's theme would apply to Agda code just the same. This is important, especially when you tend to swap themes regularly, as I do.

For this reason I have a snippet in my emacs mode like this:

```
  (progn
    (mapc
     (lambda (x) (add-to-list 'face-remapping-alist x))
     '((agda2-highlight-datatype-face              . font-lock-type-face)
       (agda2-highlight-function-face              . font-lock-type-face)
       (agda2-highlight-inductive-constructor-face . font-lock-function-name-face)
       (agda2-highlight-keyword-face               . font-lock-keyword-face)
       (agda2-highlight-module-face                . font-lock-constant-face)
       (agda2-highlight-number-face                . nil)
       (agda2-highlight-postulate-face             . font-lock-type-face)
       (agda2-highlight-primitive-type-face        . font-lock-type-face)
       (agda2-highlight-record-face                . font-lock-type-face))))
```

Which gives me good results, but can perhaps be improved if you are more familiar with the built-in faces.

**Comment by EskoJTH** [comment]
Thanks for the answers! 
In the case of rainbow-delimiters-mode it is an emacs mode and I did not mean bundling it with agda. The issue is just that agda seems to overwrite the coloring of the parenthesis. The mode has worked very well with everything else I have done before this like haskell and intero for example. I'm not quite sure how this mode sets the coloring for the parenthesis though. It seems to have color faces for different levels of nesting and then they are somehow interactively set from somewhere.

I updated my original question very slightly.

**Comment by EskoJTH** [comment]
@ajrouvoet This is exactly what I meant in regard to the color themes. This setup seems to work fine for me. I'm not very experienced with configuring the built-in themes in general as they usually have been working out out of the box for me with only little tweaking. This was one of the times when some of the dark blueish colors faded into my background very well though.

**Comment by nad** [comment]
> I think this is indeed a problem with the current setup. Agda uses non-standard classes of syntax where it could make use of the usual syntactic classes.

Have you tried to set `agda2-highlight-face-groups` to `default-faces`?

https://github.com/agda/agda/blob/02fe8918eb5c22bf75de040a4952c148858badb0/src/data/emacs-mode/agda2-highlight.el#L218-L232

**Comment by nad** [comment]
> Would It be possible to allow rainbow-delimiters mode to color parenthesis? Currenlty agda is somehow overwriting its behaviour.

By default Agda uses the face `agda2-highlight-symbol-face` to highlight parentheses in code. Does the rainbow-delimiters mode start working if you remove all attributes from this face? (You can access a customisation interface by typing `M-x customize-face RET agda2-highlight-symbol-face RET`.)

**Comment by ajrouvoet** [comment]
> Have you tried to set agda2-highlight-face-groups to default-faces?

I haven't yet, but will. It might be good to document this [here](https://agda.readthedocs.io/en/v2.6.1.1/tools/emacs-mode.html) as well.

**Comment by EskoJTH** [comment]
I did some test based on the suggestions here.

I started by testing with an Emacs with all agda related custom face settings removed from init. Rainbow-delimiters does not work.

> Have you tried to set `agda2-highlight-face-groups` to `default-faces`?

I do like the way my own colors look on a glance. Not sure if there are some errors that it doesn't show or something similar though. I'm a bit too new to Agda to test that properly. Also didn't help for rainbow-delimiters. 

> By default Agda uses the face agda2-highlight-symbol-face to highlight parentheses in code. Does the rainbow-delimiters mode start working if you remove all attributes from this face

Removing all marks from the face added the following to init:
'(agda2-highlight-symbol-face ((t nil)))
Does not help.

With both cases I went on and disabled and enabled rainbow delimiters in a freshly started emacs.
